File: exercise3.py
"""
Modeling of Reactions
Exercise 3: Freundlich sorption
s = K * c^n
m = Vw*c + ms*s = (Vw + ms*K*c^(n-1)) * c
"""

import numpy as np
import time


# Constants
n_Fr  = 0.7
K_Fr  = 2          # [mol^.3m^2.1/kg]
rho_s = 2650       # [kg/m3]
poros = 0.4

# Tolerance for mass balance
tol = 1e-7

# Mass of solids and volume of water in 1 m³ of bulk volume
ms = rho_s * (1 - poros)
Vw = poros

# Initial conditions
c0 = 10  # [mol/m³]
s0 = 0   # [mol/kg]
m  = Vw * c0 + ms * s0  # total mass [mol]

# -------------------------------
# Picard Iteration
# -------------------------------
print("Solve equilibration using Picard iteration")

c = c0
s = K_Fr * c**n_Fr
m_error = m - Vw * c - ms * s
iter_count = 0

while abs(m_error) > tol * m:
    iter_count += 1
    c = m / (Vw + ms * K_Fr * c**(n_Fr - 1))
    s = K_Fr * c**n_Fr
    m_error = m - Vw * c - ms * s
    print(f"iteration {iter_count:2d}: c = {c:8.3g} mol/m³, error = {m_error:8.3g} mol")

print(f"sorbing-phase concentration {s:.6g} mol/kg")

# -------------------------------
# Newton Iteration
# -------------------------------
print("Solve equilibration using Newton iteration")

c = c0
iter_count = 0
s = K_Fr * c**n_Fr
res = m - Vw * c - ms * s

while abs(res) > tol * m:
    iter_count += 1
    dresdc = -Vw - ms * K_Fr * n_Fr * c**(n_Fr - 1)
    dc = -res / dresdc

    # Line search to keep concentration non-negative
    fac = 1.0
    while c + fac * dc < 0:
        fac *= 0.9

    c += fac * dc
    s = K_Fr * c**n_Fr
    res = m - Vw * c - ms * s
    print(f"iteration {iter_count:2d}: c = {c:8.3g} mol/m³, error = {res:8.3g} mol")

print(f"sorbing-phase concentration {s:.6g} mol/kg")

File: exercise3.m
% Modeling of Reactions
% Exercise 3: Freundlich sorption
% s = K*c^n
% total mass:
% m = Vw*c + ms*s = (Vw + ms*K*c^(n-1))*c
% Picard iteration:
% c_new = m/(Vw + ms*K*c_old^(n-1))
clear variables
close all
clc

% coefficients
n_Fr  = 0.7;
K_Fr  = 2;    % [mol^.3m^2.1/kg]
rho_s = 2650; % [kg/m3]
poros = 0.4;

% relative mass-balance error
tol=1e-7;

% mass of solids and volume of water in 1m3 of bulk volume
ms = rho_s*(1-poros);
Vw = poros;

% total mass from initial concentration
c0 = 10; % [mol/m3]
s0 = 0;  % [mol/kg]
m  = Vw*c0 + ms*s0; % [mol]

%% Picard iteration
disp('Solve equilibration using Picard iteration')
tic
% iteration
c = c0;
s = c^n_Fr*K_Fr;
m_error = m-Vw*c-ms*s;
iter = 0;
while abs(m_error)>tol*m
      iter=iter+1;
      c = m/(Vw + ms*K_Fr*c^(n_Fr-1));
      s = K_Fr*c^n_Fr;
      m_error = m-Vw*c-ms*s;
      fprintf('iteration %2.2i: c = %8.3g mol/m3, error = %8.3g mol\n',...
              [iter,c,m_error])
end
fprintf('sorbing-phase concentration %g mol/kg\n',s)
toc

%% Newton iteration
disp('Solve equilibration using Newton iteration')
tic
% iteration
c = c0;
iter = 0;
s = K_Fr*c^n_Fr;
res = m-Vw*c-ms*s;
while abs(res)>tol*m
      iter=iter+1;
      dresdc = -Vw -ms*K_Fr*n_Fr*c^(n_Fr-1);
      dc = - res/dresdc;
      fac= 1;
      while c+fac*dc < 0
         fac = fac*.9;
      end
      c = c + fac*dc;
      s = K_Fr*c^n_Fr;
      res = m - Vw*c - ms*s;
      fprintf('iteration %2.2i: c = %8.3g mol/m3, error = %8.3g mol\n',...
              [iter,c,res])
end
fprintf('sorbing-phase concentration %g mol/kg\n',s)
toc
